### PR TITLE
Adding NCHWxC and KCMNxCyK data layout to ONNX dialect

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Added the data layout that will be helpful to the optimizations of convolutions.

Right now, the data layout explicitly encode the x and the y unroll factor in the MLIR textual representation, but internally the factors are treated as a parameter of the ONNX Tensor Encoding attributes.

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
@tungld since you did this one for `zhigh`, do you mind having a look if it looks about right? Tx

File: src/Dialect/ONNX/ONNX.td [hunk 1]
```diff
@@ -54,6 +54,105 @@ def ONNX_Dialect : Dialect {
...
+    with only one tiling dimension, `y` should be zero.
+
+    Example:
+
+    ```mlir
+    #ONNX_NCHWxC= #onnx.encoding<{
+      dataLayout = "NCHWxC",
+      xFactor = 4,
+      yFactor = 0
+    }>
```
Comment (tungld) [inline src/Dialect/ONNX/ONNX.td:87]: @AlexandreEichenberger yes, this looks right to me. Thanks!

**Comment by AlexandreEichenberger** [comment]
@tungld ready for review. I will use this in the next PR, simply introducing CONV with this layout and 2 copy from/to operators.

**Review by tungld — APPROVED** [review]
LGTM!

File: src/Dialect/ONNX/ONNX.td [hunk 2]
```diff
@@ -54,6 +54,105 @@ def ONNX_Dialect : Dialect {
...
+// Whether a ztensor type has the specified layout. 
+//class ONNXCustomDataLayoutOfPred<string layout> :
+// And<[
+//  CPred<"($_self.cast<::mlir::RankedTensorType>()) &&"
+//        "($_self.cast<::mlir::RankedTensorType>()."
+//        "getEncoding().dyn_cast_or_null<mlir::ONNXTensorEncodingAttr>()) &&"
+//        "($_self.cast<::mlir::RankedTensorType>()."
+//        "getEncoding().cast<ONNXTensorEncodingAttr>().getDataLayout()"
+//        " == ONNXTensorEncodingAttr::DataLayout::" # layout # ")"> 
+//]>;
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Dialect/ONNX/ONNX.td:107]: yes, tx

**Comment by AlexandreEichenberger** [comment]
@philass The PR is ready to go but will leave open for a bit for you to go over it and provide feedback if you want to. As discussed yesterday, it could be part of a different dialect, but then it may eventually require a lot of redundant ops (esp for the activation/ops that are typically found between convolutions). So the idea was to enable a pass that convert normal layout to custom layout only on a specific option, and without it, all layout would remain the traditional layouts. I think this approach may also help others if they have backends that need different layouts, as the mechanics to introduce them and optimize them can become shared.

**Comment by philass** [comment]
@AlexandreEichenberger I definitely understand the desire to capture this information at the `onnx` dialect level. it can be a lot easier to reason about and write certain optimizations at this higher level of abstraction. 

In the Groq code base we have a Dialect called ONNX extension ops. We rewrite some ONNX ops into these to capture things that we don't think are modeled well in `ONNX` or for things we need more Groq specific representations.

A reason it might make sense to have this as a standalone op is the following

The op definitions are generated using the python script. This means that although most of the information is redundant with the pre-existing ONNX Conv ops, its not too painful as the tablegen generating script will take care of that for us. 

The pass that introduces these attributes to the conv op could instead rewrite into the modified conv op (the new one with its own definition).

An example I can maybe share is that we don't have native support for I64 in our hardware. We debated heavily in our fork about whether we should be relaxing the type restrictions of the onnx argmax by modifying it to accept `i32` as an additional output type. We ultimately decided against it because we felt that it was the wrong level of abstraction to do that. We felt that creating a new op ArgMaxRelaxed that allowed for ArgMax to have more relaxed output types was more correct.

However it going in as its currently written won't cause problems. I would just argue that it leaks a hardware/implementation specific abstraction in the wrong place. From an implementations standpoint I can see why this is desireable (less code reuse). I imagine creating a new op would also require adding shape inference and finding ways to reuse alot of the codepaths that onnx conv ops has.

I've said a lot. In short, I have a mild preference for adding a new op to capture this. But if this was going to go in as is thats totally fine too!

**Comment by tungld** [comment]
```
 'Conv': [{0 : ['ONNXTensor_NCHW4C'], 1 : ['ONNXTensor_KCMN4C4K']},
              {0 : ['ONNXTensor_NCHW4C']}]
```

Have you checked if it works without adding these new types? I though that `TensorOf<f32>` does not check TensorEncoding attribute so it already covers `ONNXTensor_NCHW4C`, `ONNXTensor_KCMN4C4K`, `ONNXTensor_NCHW4C`.

**Comment by AlexandreEichenberger** [comment]
> Have you checked if it works without adding these new types

good idea, let me try

**Comment by chentong319** [comment]
I think the new type should be only added to ZHigh or ZLow Dialect, not in ONNX dialect.

**Comment by AlexandreEichenberger** [comment]
> I think the new type should be only added to ZHigh or ZLow Dialect, not in ONNX dialect.

High/low are for targeting NNPA. This PR targets the CPU code

**Comment by AlexandreEichenberger** [comment]
> Have you checked if it works without adding these new types

it works without the new types... have a new PR that supersede this one.

**Comment by AlexandreEichenberger** [comment]
new pr is #1762

**Pull request closed without merging**